In [2]:
import pandas as pd
import numpy as np

## Part 1: Data Loading and Exploration

In [3]:
# Load the California Housing Dataset

from sklearn.datasets import fetch_california_housing

# Load the housing dataset
housing = fetch_california_housing()

In [ ]:
# Creating the inital dataframe for the features of the dataset and Series for 

X = pd.DataFrame(housing.data, columns = housing.feature_names) 
y = pd.Series(housing.target, name = 'med_house_value')

In [32]:
X.head() # displays the first five rows of the dataset

MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0  8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1  8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2  7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3  5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4  3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   

   Longitude  
0    -122.23  
1    -122.22  
2    -122.24  
3    -122.25  
4    -122.25

In [35]:
print("Feature names:", housing.feature_names) # prints the column names within the "feature names"


Feature names: ['MedInc', 'HouseAge', 'AveRooms', 'AveBedrms', 'Population', 'AveOccup', 'Latitude', 'Longitude']


In [43]:
print("Amount of missing values in the feature names:\n", X.isnull().sum()) # .isnull() puts boolean mask on the entire dataframe; 
# .sum() adds up how many nulls you have in a column

Amount of missing values in the feature names:
 MedInc        0
HouseAge      0
AveRooms      0
AveBedrms     0
Population    0
AveOccup      0
Latitude      0
Longitude     0
dtype: int64


In [ ]:
X.describe() # gives summary statistics of the dataframe

MedInc      HouseAge      AveRooms     AveBedrms    Population  \
count  20640.000000  20640.000000  20640.000000  20640.000000  20640.000000   
mean       3.870671     28.639486      5.429000      1.096675   1425.476744   
std        1.899822     12.585558      2.474173      0.473911   1132.462122   
min        0.499900      1.000000      0.846154      0.333333      3.000000   
25%        2.563400     18.000000      4.440716      1.006079    787.000000   
50%        3.534800     29.000000      5.229129      1.048780   1166.000000   
75%        4.743250     37.000000      6.052381      1.099526   1725.000000   
max       15.000100     52.000000    141.909091     34.066667  35682.000000   

           AveOccup      Latitude     Longitude  
count  20640.000000  20640.000000  20640.000000  
mean       3.070655     35.631861   -119.569704  
std       10.386050      2.135952      2.003532  
min        0.692308     32.540000   -124.350000  
25%        2.429741     33.930000   -121.800000  
50%        2.818116     34.260000   -118.490000  
75%        3.282261     37.710000   -118.010000  
max     1243.333333     41.950000   -114.310000

## Part 2: Linear Regression on Unscaled Data

In [44]:
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split

In [ ]:
# splitting the dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, # this randomly splits the data (80% training, 20% testing)
                                                    test_size = 0.2,
                                                    random_state = 6)
# check to see if it is split correctly
print(y_train.shape)
print(y_test.shape)

(16512,)
(4128,)


In [ ]:
# train the linear regression model
lin_reg = LinearRegression()
lin_reg.fit(X_train, y_train)
# if a bubble with the linear regression shows up it means we did it right!

LinearRegression()

In [48]:
# making predictions on the test set
y_pred = lin_reg.predict(X_test)
y_pred

array([2.33035401, 2.7322844 , 3.33997869, ..., 2.25209662, 2.49516329,
       1.51341588])

In [ ]:
# evaluating model performance
from sklearn.metrics import mean_squared_error, root_mean_squared_error, r2_score
mse = mean_squared_error(y_test, y_pred) # mean squared error measures the average gap between the predicted values and the values in the data
print("Mean Squared Error:", mse)
rmse = root_mean_squared_error(y_test, y_pred) # root mean squared error is the squared root of MSE; this accounts for negative gaps
print("Root Mean Squared Error:", rmse)
r2 = r2_score(y_test, y_pred) # the R-squared is also called the coefficient of determination and tells us how well the model explains variance in the data
print("R-squared:", r2)

Mean Squared Error: 0.5384002833539473
Root Mean Squared Error: 0.733757646197944
R-squared: 0.5947275061211218


In [55]:
# seeing the coefficients of our model
coefficients = pd.Series(lin_reg.coef_, index = X.columns)
coefficients

MedInc        4.403567e-01
HouseAge      9.840493e-03
AveRooms     -1.125770e-01
AveBedrms     6.529406e-01
Population   -2.454707e-07
AveOccup     -4.102497e-03
Latitude     -4.161599e-01
Longitude    -4.299264e-01
dtype: float64

1. What does the R² score tell us about model performance? 

The R² score tells us about how well our linear regression model explains the variation in the target that we chose, which would be "med_house_value." Another way of looking at it simply is that the R² tells us how well our model fits the data. If we had an R² of 1.0, it would mean that our model can perfectly predit the values that are in the dataset and a value of 0.0 means that the model explains none of the variance in the data. We have an R² of 0.5384, which means that our model can account for about 54% of the variance in the data. Although this is not bad, there is still 48% of variance left unexplained, which could indicate there's other important factors affecting house prices that we haven't yet accounted for.

2. Which features seem to have the strongest impact on predictions based on the model’s coefficients?

The bigger in magnitude the coefficient is, the most impact or weight each feature has on our prediction of y. The most significant coefficients I can tell are Median Income, Average Bedrooms, and Latitude and Longitude. Based on the coefficients, we can see that as median income increases by one unit, the predicted house price increases by 0.44. Similarly, when the average number of bedrooms incraeses by one unit, the predicted house price increases by .65. The magnitude of these coefficients is not too large, however, which could mean that there is a factor affecting house prices that we have not included in our analysis yet.

3. How well do the predicted values match the actual values?

I would say that an evaluation of our model and how well it fits the data shows us that the predicted values are not a great match for the actual ones. As we have discussed in question 1, the R² we have is not very high and leaves us with almost half of the variance in the data unexplained. Additionally, we can look at the MSE adn RMSE and see that both values are pretty high, with the RMSE being 0.73. An RMSE of 0.73 means that on average, the model's predictions are off from the actual data by about 0.73 units. If, for example, the unit of measurement was $100,000, this would mean that our model, on average, is off by about $73,000, which is significant. Furthermore, we can see from looking at our coefficients that although some are stronger in magnitude than others, none of the coeffients are exceptionally large and show that they do not have a significant impact on our predictions. Therefore, the predicted values do not match the actual values very well.  

## Part 4: Feature Selection and Simplified Model

The three features I want to select are those that are most impactful for housing prices and also interesting to look at. These features are: MedInc, AveBedrms, and HouseAge. Based off of the coeffients from earlier and life experience, I'd expect Median Income to be significant in the prices of houses, houses with greater number of bedrooms being more expensive, and houses that are much older being cheaper than those that are newly built.  

In [57]:
# Creating the simplified model
selected_features = ["MedInc", "AveBedrms", "HouseAge"]
simplified_model = X[selected_features]
simplified_model

MedInc  AveBedrms  HouseAge
0      8.3252   1.023810      41.0
1      8.3014   0.971880      21.0
2      7.2574   1.073446      52.0
3      5.6431   1.073059      52.0
4      3.8462   1.081081      52.0
...       ...        ...       ...
20635  1.5603   1.133333      25.0
20636  2.5568   1.315789      18.0
20637  1.7000   1.120092      17.0
20638  1.8672   1.171920      18.0
20639  2.3886   1.162264      16.0

[20640 rows x 3 columns]

In [ ]:
# Training the new linear regression model
X_train_simplified, X_test_simplified, y_train, y_test = train_test_split(X, y, # splitting the dataset
                                                                          test_size = 0.2,
                                                                          random_state = 41)
simplified_model = LinearRegression()
simplified_model.fit(X_train_simplified, y_train) # training the model

y_pred_simplified = simplified_model.predict(X_test_simplified) # making predictions
y_pred_simplified

array([1.42990818, 2.33445234, 1.74646549, ..., 0.46987616, 2.87541556,
       0.75920043])

In [59]:
# Evaluating performance
print("Simplified Model Performance:")
mse_simplified = mean_squared_error(y_test, y_pred_simplified)
print("Mean Squared Error:", mse_simplified)
rmse_simplified = root_mean_squared_error(y_test, y_pred_simplified)
print("Root Mean Squared Error:", rmse_simplified)
r2_simplified = r2_score(y_test, y_pred_simplified)
print("R-squared:", r2_simplified)

Simplified Model Performance:
Mean Squared Error: 0.5325748833763695
Root Mean Squared Error: 0.729777283406636
R-squared: 0.5950798933780448


In [60]:
print("Full Model Performance:")
print("Mean Squared Error:", mse)
print("Root Mean Squared Error:", rmse)
print("R-squared:", r2)

Full Model Performance:
Mean Squared Error: 0.5384002833539473
Root Mean Squared Error: 0.733757646197944
R-squared: 0.5947275061211218


1. How does the simplified model compare to the full model?

The simplified model is almost exactly the same as the full model. Except for changes in the thousandths decimal place in the performance metrics, there is no difference at all between the MSE, RMSE, or R² in the simplified model than what we had already seen in the full model.

2. Would you use this simplified model in practice? Why or why not?

I would use the simplified model in practice over the full model. Even though there was almost no difference at all between the evaluative metrics between both models, this is precisely why using a much simpler model would be easier to interpret, more efficient, and faster to train. The R² being almost exactly the same tells us that the other features in the full model must not have been very influencial or had much predictive power, therefore it's not an issue if we drop those features in favor of a simpler model.